sdcclient/_scanning.py

```diff
@@ -0,0 +1,535 @@
+        '''
+        return self._update_subscription(subscription_type, subscription_key, False)
+
+    def _update_subscription(self, subscription_type, subscription_key, activate):
+        hashstr = '+'.join([self.token, subscription_key, subscription_type]).encode('utf-8')
+        subscription_id = hashlib.md5(hashstr).hexdigest()
+        url = self.url + "/api/scanning/v1/anchore/subscriptions/" + subscription_id
+        payload = {'active': activate, 'subscription_key': subscription_key, 'subscription_type': subscription_type}
+
+        res = requests.put(url, data=json.dumps(payload), headers=self.hdrs, verify=self.ssl_verify)
+        if not self._checkResponse(res):
+            return [False, self.lasterr]
+
+        return [True, res.json()]
+
+    def list_subscription(self):
+        '''**Description**
+            List all subscriptions
+
+        **Arguments**
+            - None
+
+        **Success Return Value**
+            A JSON object representing the list of subscriptions.
+        '''
+        url = self.url + "/api/scanning/v1/anchore/subscriptions"
+        res = requests.get(url, headers=self.hdrs, verify=self.ssl_verify)
+        if not self._checkResponse(res):
+            return [False, self.lasterr]
+
+        return [True, res.json()]
+
+    def _discover_inputimage_format(self, input_string):
+        itype = None
+
+        if re.match("^sha256:[0-9a-fA-F]{64}", input_string):
+            itype = 'imageDigest'
+        elif re.match("[0-9a-fA-F]{64}", input_string):
+            itype = 'imageid'
+        else:
+            itype = 'tag'
+
+        return(itype)
+
+    def _discover_inputimage(self, input_string):
+        patt = re.match(".*(sha256:.*)", input_string)
+        if patt:
+            urldigest = quote_plus(patt.group(1))
+            return("digest", input_string, urldigest)
+
+        try:
+            digest = unquote_plus(str(input_string))
+            patt = re.match(".*(sha256:.*)", digest)
+            if patt:
+                return("imageDigest", input_string, input_string)
+            patt = re.match(".*(local:.*)", digest)
+            if patt:
+                return("imageDigest", input_string, input_string)
+        except Exception:
+            pass
+
+        urldigest = None
+        ret_type = "tag"
+        try:
+            ok, ret = self.get_image(input_string)
+            if ok:
+                urldigest = ret[0]['imageDigest']
+                try:
+                    image_record = ret[0]
+                    for image_detail in image_record['image_detail']:
+                        if input_string == image_detail['imageId']:
+                            ret_type = "imageid"
+                            break
+                except Exception:
+                    pass
+            else:
+                pass
+        except Exception:
+            urldigest = None
```

Comment:
What exception are we capturing here? Dict keys not found?
We can then simplify code:
```python
ret_type = "tag"
urldigest = None
ok, ret = self.get_image(input_string)
if ok:
    image_record = ret[0]
    urldigest = image_record.get('imageDigest', None)
    for image_detail in image_record.get('image_detail', []):
        if input_string == image_detail.get('imageId', ''):
            ret_type = "imageid"
            break
return(ret_type, input_string, urldigest)
```

Reply:
I'm not even sure what exception is. Most of this code is kind of copy-paste from anchore-cli, I should had cleaned it up more.